### PR TITLE
feat(job_shop): implement dense schedule generator

### LINKED ISSUE #90: feat(jobshop): create instance generator with known optimal solution
### Is your feature request related to a problem? Please describe
Create a `JobShop` instance generator whose optimal makespan is known. This will be useful to benchmark agents better since the best solution will be known in advance. 

### Describe the solution you'd like
The generator will generate a random schedule based on a specified makespan (length of the schedule), number of machines, number of jobs, and max number of operations per job.

### PR DESCRIPTION
**Pull request by dluo96**
Closes #90. 

- Implemented the `DenseGenerator`, which generates an instance of the job shop scheduling problem with a _specified_ schedule length (aka makespan). This is useful for benchmarking RL agents because the optimal return is known in advance. 
- Wrote unit tests for the generator. 
- Tested this generator with the job shop A2C agent.

### CODE REVIEW

**Comment by clement-bonnet** [comment]
Hi @dluo96, thanks for the contribution! Can you please update the default generator to this new one, and update the docstring, registration and readme accordingly?

File: jumanji/environments/packing/job_shop/generator_test.py [hunk 1]
```diff
@@ -103,3 +105,101 @@ def test_random_generator__call(self, random_generator: RandomGenerator) -> None
...
+        state = dense_schedule_generator(key)
+        assert isinstance(state, State)
+
+    def test_dense_schedule_generator__generate_schedule(
+        self, dense_schedule_generator: DenseGenerator
+    ) -> None:
+        key = jax.random.PRNGKey(0)
+        schedule = dense_schedule_generator._generate_schedule(key)
+        assert schedule.shape == (self.NUM_MACHINES, self.MAKESPAN)
+        assert jnp.all((schedule >= 0) & (schedule < self.NUM_JOBS))
```
Comment (coyettev) [inline jumanji/environments/packing/job_shop/generator_test.py:157]: Maybe we could add a check that no column contains twice the same value ? something like
```python
        _, counts = jnp.unique(schedule, return_counts=True, axis=0)
        assert jnp.all(counts <= 1)
```

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Thanks a lot for this generator. Could we refactor the code to reduce the amount of nesting? Also, I suggest adding the generator as a possible option but not defaulting to it in the env init. We can default to it and up the version in another PR after testing a complete training with it (and potentially adapt the network). What do you think?

File: jumanji/environments/packing/job_shop/generator.py [hunk 2]
```diff
@@ -184,3 +185,297 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+        Args:
+            num_jobs: the number of jobs that need to be scheduled.
+            num_machines: the number of machines that the jobs can be scheduled on.
+            max_num_ops: the maximum number of operations for any given job.
+            max_op_duration: the maximum processing time of any given operation.
+            makespan: the length of the schedule. By construction, this will be the
+                shortest possible length of the schedule.
+        """
+        del max_op_duration
+        del max_num_ops
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator.py:228]: Sounds good, have updated the doc string - let me know if you're happy with that

File: jumanji/environments/packing/job_shop/generator_test.py [hunk 3]
```diff
@@ -103,3 +105,101 @@ def test_random_generator__call(self, random_generator: RandomGenerator) -> None
 
         state2 = call_fn(key=jax.random.PRNGKey(2))
         assert_trees_are_different(state1, state2)
+
+
+class TestMakespanGenerator:
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator_test.py:110]: Agreed, have done 👍

File: jumanji/environments/packing/job_shop/generator_test.py (hunk 1 above)
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator_test.py:157]: Good idea, have added this check to the test 👍

File: jumanji/environments/packing/job_shop/generator.py [hunk 4]
```diff
@@ -184,3 +185,297 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+                        None,
+                    )
+                    _job_mask = _job_mask.at[job_id].set(False)
+                    return (_key, _job_mask, _machine_id + 1), job_id
+
+                # Define initial conditions for the scan
+                # init_job_mask = jax.vmap(
+                #     lambda _job_id, _col: ~jnp.any(_col == _job_id),
+                #     in_axes=(0, None),
+                # )(all_job_ids, prev_col)
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator.py:334]: Yup, this was just an alternative way to set the initial job mask. Will remove 👍

File: jumanji/environments/packing/job_shop/generator.py [hunk 5]
```diff
@@ -184,3 +185,297 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+    This generator assumes that the number of jobs is less than or equal to the number of
+    machines.
+    """
+
+    def __init__(
+        self,
+        num_jobs: int,
+        num_machines: int,
+        max_num_ops: int,
+        max_op_duration: int,
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator.py:213]: With the way that the `Generator` interface is currently defined, we would have to, yes. This is because the `__init__` takes those arguments.

File: jumanji/environments/packing/job_shop/generator.py [hunk 6]
```diff
@@ -184,3 +185,297 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+            carry: Tuple[chex.PRNGKey, int, chex.Array], _: Any
+        ) -> Tuple[Tuple[chex.PRNGKey, int, chex.Array], chex.Array]:
+            """Create a column of distinct job ids (only one operation in a given job
+            can be processed at a time). For the example above, this would be [1, 4, 0] at
+            time=0, [0, 1, 2] at time=1, etc.
+            """
+            key, t, prev_col = carry
+            key, job_key, reuse_key = jax.random.split(key, num=3)
+
+            def reuse_prev_col(key: chex.PRNGKey, prev_col: chex.Array) -> chex.Array:
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator.py:307]: I tried this but unfortunately I get some JAX tracing errors on this line:
```
job_mask = jnp.ones_like(all_job_ids).at[prev_col[jnp.where(reuse_jobs)]].set(0)
```

File: jumanji/environments/packing/job_shop/generator.py [hunk 7]
```diff
@@ -184,3 +185,297 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+
+                return col
+
+            def sample_new_jobs(key: chex.PRNGKey) -> chex.Array:
+                return jax.random.choice(
+                    key, all_job_ids, (self.num_machines,), replace=False
+                )
+
+            # Reuse the previous column with probability 0.6
+            reuse = jax.random.uniform(reuse_key, shape=()) > 0.4
```
[the remark this answers is not in the document]
Comment (dluo96) [reply jumanji/environments/packing/job_shop/generator.py:353]: Yeah I reckon that's a good idea

**Comment by clement-bonnet** [comment]
Closing this for now, will re-open after v0.3
